blog/2024-05-21-blog-status-of-lbse-in-webkit.md

```diff
@@ -0,0 +1,149 @@
+
+As the first SVG resource to implement we chose the SVG &lt;clip-path&gt;, so we landed <a href="https://commits.webkit.org/269635@main"><code>RenderSVGResourceClipper</code></a>.
+
+To comply with the specification, the `RenderSVGResourceClipper` implementation produces 1-bit masks and uses a special rendering mode:
+
+- fill-opacity/stroke-opacity/opacity set to 1
+- masker/filter not applied when rendering the children
+- fill set to solid black and stroke set to none
+
+The initial implementation did not use caching of ImageBuffers for clipping, but relied on Porter-Duff DestinationIn/SourceOver compositing operations, to achieve the same effect, but faster. By integrating `RenderSVGResourceClipper` properly into `RenderLayer`, it aligned SVG clipping with HTML/CSS clipping.
+
+Finally, internally the implementation (like in legacy), prefers a pure clipping solution, but for more complicated clip-path's (for example when the clip-path involves text content) a fallback to a mask is done.
+
+## Resource invalidation handling
+
+After introducing the first SVG resource (`RenderSVGResourceClipper`) we noticed some issues with handling invalidations (for example adding to clip-path contents) for it. In the legacy engine, invalidations have been handled through layouting. This caused various problems, for one it could cause calling the setNeedsLayout method from within layout and the invalidation chain depended on the DOM order.
+
+In **November 2023** an <a href="https://commits.webkit.org/271017@main">implementation</a> landed that avoided using layout for resource invalidation. Instead, on dynamic updates the style system is used to determine the appropriate action:
+
+- For non resources, changes that cause renderer geometry changes, like changing the x value of a &lt;rect&gt; element still require a relayout. For visual changes not affecting geometry, like changing fill color on a &lt;rect&gt;, a repaint action is enough.
+- For resources, the resource is invalidated/updated with the change and any of its clients are repainted using the new resource.
+
+## Support for masks
+
+With improved support for SVG resource invalidation, in **late November 2023** we were ready to upstream support for the next SVG resource, <a href="https://commits.webkit.org/271153@main">RenderSVGResourceMasker</a>.
+
+Like the support for clip-path, it started out without caching image buffers and makes use of Porter-Duff DestinationIn/SourceOver compositing to efficiently render the mask.
```

Comment:
We always used ImageBuffers for masking, just uncached initially. No Porter/Duff here.

Reply:
Good catch, it is easy to mix these up, fixed.